python_ta/contracts/__init__.py

```diff
@@ -1,40 +1,70 @@
 from typing import Any, List, Optional
 import typing
+import inspect
 import wrapt
 
 
-class InvalidPrecondition(Exception):
-    """Exception raised when precondition is invalid
+class InvalidCondition(Exception):
+    """Exception raised when a condition is invalid
     """
 
+
 @wrapt.decorator
-def check_contracts(wrapped, _instance, args, kwargs):
+def check_contracts(wrapped, instance, args, kwargs):
+    if instance is None and inspect.isclass(wrapped):
+        return add_class_contracts(wrapped, args, kwargs)
+    return check_function_contracts(wrapped, instance, args, kwargs)
+
+
+def add_class_contracts(wrapped, args, kwargs):
+    rep_invariants = parse_conditions(
+        wrapped.__doc__ or '', parse_token="Representation Invariant")
+    setattr(wrapped, "__representation_invariants__", rep_invariants)
+    return wrapped(*args, **kwargs)
+
+
+def check_function_contracts(wrapped, instance, args, kwargs):
     params = wrapped.__code__.co_varnames[:wrapped.__code__.co_argcount]
     annotations = typing.get_type_hints(wrapped)
-    for arg, param in zip(args, params):
+
+    args_with_self = (instance,) + args if instance else args
+
+    for arg, param in zip(args_with_self, params):
         if param in annotations:
             assert check_type_annotation(annotations[param], arg),\
                 f'Argument {repr(arg)} did not match type annotation for parameter "{param}: {annotations[param]}"'
 
-    preconditions = parse_preconditions(wrapped.__doc__ or '')
-    function_locals = dict(zip(params, args))
-    for precondition in preconditions:
-        try:
-            check = eval(precondition, wrapped.__globals__, function_locals)
-        except:
-            # TODO: Decide what to do here, e.g. "Invalid precondition"
-            raise InvalidPrecondition(
-                f'Error evaluating precondition: {precondition}')
-        else:
-            assert check,\
-                f'Precondition "{precondition}" violated for arguments {function_locals}'
+    preconditions = parse_conditions(wrapped.__doc__ or '')
+
+    function_locals = dict(zip(params, args_with_self))
+    validate_conditions(wrapped, function_locals, preconditions)
+
     r = wrapped(*args, **kwargs)
+
+    if instance and hasattr(type(instance), "__representation_invariants__"):
+        validate_conditions(wrapped, function_locals,
+                            type(instance).__representation_invariants__)
+
     if 'return' in annotations:
         return_type = annotations['return']
         assert check_type_annotation(return_type, r),\
             f'Return value {r} does not match annotated return type {return_type.__name__}'
     return r
 
+
+def validate_conditions(wrapped, function_locals, conditions):
```

Comment:
In addition to swapping out "conditions" for "assertions", I worry that the use of "validate" here makes it sound like this function checks whether the assertions are formed properly, rather than whether they're actually satisfied by the given `function_locals`.

Let's stick with "check", i.e., `def check_assertions` as the header here. This is a generic enough name that it can still be used for both preconditions and representation invariants.

Reply:
That makes sense. I've changed it.